### ISSUE TITLE
I want to know how to deal with label confusion when there are many nodes.

### ISSUE THREAD

**Issue by wxxwjef** [HelikarLab/ccNetViz#128]
![微信图片_20190912153906](https://user-images.githubusercontent.com/37828799/64763912-8e7ca700-d573-11e9-8dae-7c106504e0a9.png)
And Many nodes display a different label than the actual label.

**Comment by wxxwjef** [comment]
> 你好@wxxwjef，
> 这对我来说就像第一眼看到的错误一样。
> 你能和我分享一下它失败的示例代码吗？
> 
> 谢谢，
> 阿莱斯

<!DOCTYPE html>
<html>
<head>
    <title>Bidirectional graphs example</title>
    <link rel="stylesheet" type="text/css" href="css/template.css"/>
    <script src="../lib/ccNetViz.js"></script>
</head>

<body>
<div class="canvas-container">
    <h3 class="title">Bidirectional graphs example</h3>
    <canvas id="canvas"></canvas>
    <div class="description">
        To enter to deeper node, click on node "Hello" or "World". Then use
        right button on mouse for going back.
        <br/>
        More detailed information please visit the
        <a href="https://helikarlab.github.io/ccNetViz/#doc">documentation</a>.
    </div>
</div>
<header id="logo">
    <a href="https://helikarlab.github.io/ccNetViz/">
        <img src="images/logo.svg"/>
    </a>
</header>
<script>
    var canvas = document.getElementById('canvas');
    var graph = new ccNetViz(canvas, {
        bidirectional: 'overlap',
        styles: {
            node: {texture: 'images/node.png', label: {hideSize: 16}},
            edge: {arrow: {texture: 'images/arrow.png'}},
        },
    });
    var nodes = [{label: 'Hello'}, {label: 'World'}, {label: '!'}];
    for (let i = 0; i < 2000; i++) {
        var arr = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'];
        let str = '';
        for(let j=0;j< 8;j++) {
            str += arr[Math.floor(Math.random()*26)];
        }
        nodes.push({label: str});
    }
    var edges = [
        {source: nodes[0], target: nodes[1]},
        {source: nodes[1], target: nodes[0]},
        {source: nodes[0], target: nodes[0]},
        {source: nodes[1], target: nodes[2]},
    ];
    graph.set(nodes, edges, 'force').then(() => {
        graph.draw();
    });
</script>
</body>
</html>

**Comment by wxxwjef** [comment]
Hello @saskaale 
First, thank you for your reply.I'm surprised by your reply.

The following actions can reproduce the bug.
Insert this code into line 38 of examples/bidirectional.html from official project.
My purpose is to create 2,000 nodes，when there are many nodes on the canvas, label will be confused.


// code start 
for (let i = 0; i < 2000; i++) {
        var arr = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'];
        let str = '';
        for(let j=0;j< 8;j++) {
            str += arr[Math.floor(Math.random()*26)];
        }
        nodes.push({label: str});
    }
// code end
![image](https://user-images.githubusercontent.com/37828799/64876016-d63d2480-d680-11e9-8024-7a90892e5846.png)
![image](https://user-images.githubusercontent.com/37828799/64876075-f371f300-d680-11e9-803a-5a79c3560717.png)

**Comment by canerakdas** [comment]
Hello @wxxwjef ,

Thank you for the sharing steps to reproduce.

Hey @saskaale ,
I have tested and IMHO this bug reason is static sdf canvas size (1024x1024) for the labels. We need to update dynamically.

You can see the problem with the following image;
![map](https://user-images.githubusercontent.com/14062599/64938571-09410d00-d867-11e9-9f81-4bfbfe8d6f80.png)

**Comment by wxxwjef** [comment]
@canerakdas 
Thank you for your participation.This issue has been bothering me too long.

**Comment by saskaale** [comment]
@wxxwjef thank you for the code example! Yes, this is definitely a issue in our codebase.

@canerakdas weird. Yeah, the texture seems not to be caching the Glyphs per characted and instead of it is rendering it again and again.

The caching code:
https://github.com/HelikarLab/ccNetViz/blob/master/src/texts/sdf/sdf.js#L263-L266

**Comment by saskaale** [comment]
@canerakdas Hello Caner, I just looked at it and it seems to be bigger issue overally. 
The issue is that the ccNetViz is currently using the text default generator which has the static texture size (https://github.com/HelikarLab/ccNetViz/blob/master/src/texts/default.js) - fixing of this size to be dynamiccaly resized seems like a solution. 


However we tried to consider the default text generator as a deprecated so we are would be using the SDF one which is supposed to have better performance. It would also not have any issue with too many nodes with long text because it works character-wise intead of whole label and also internally supports lazy character caching. 
To enable the SDF you can replace these 3 lines (https://github.com/HelikarLab/ccNetViz/blob/master/src/texts/texts.js#L30-L35) with the return true; . Sadly once you do it, the library does not show anything at all. 
- that is my investigation so far - I am looking at it currently more deeply.


thanks,
Ales

**Comment by hjhimanshu01** [comment]
Hey @wxxwjef , @saskaale , I'm not sure how to fix default font size , but as default is basically deprecated and @wxxwjef , just change this part in the code
```
styles: {
          node: { texture: 'images/node.png', label: { hideSize: 16 , font:{ type: "sdf", size: 10}} },
          edge: { arrow: { texture: 'images/arrow.png' } },
        }
```
basically the size of label is dynamic and on 2000 nodes, size becomes minimal, so changing the font type and size does the work.

**Comment by wxxwjef** [comment]
@hjhimanshu01 
Thank you for your proposal.
I found the performance of this kind of configure is very poor.How to improve performance.

**Comment by saskaale** [comment]
@saskaale do you mean poor in terms of the performance (rendering speed) ?

**Comment by wxxwjef** [comment]
@saskaale 
Yes,I tested it locally with type of 'sdf',found the process of Initialization is very slow when graph hava a lot nodes.
